external/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/CachedKafkaProducerSuite.scala

```diff
@@ -18,60 +18,206 @@
+      "value.serializer" -> classOf[ByteArraySerializer].getName)
 
-    CachedKafkaProducer.close(kafkaParams)
-    val map2 = CachedKafkaProducer.invokePrivate(cacheMap())
-    assert(map2.size == 1)
     import scala.collection.JavaConverters._
-    val (seq: Seq[(String, Object)], _producer: KP) = map2.asScala.toArray.apply(0)
-    assert(_producer == producer)
+
+    val numThreads: Int = 100
+    val numConcurrentProducers: Int = 1000
+
+    val kafkaParamsUniqueMap = mutable.HashMap.empty[Int, ju.Map[String, Object]]
+    ( 1 to numConcurrentProducers).map {
+      i => kafkaParamsUniqueMap.put(i, kafkaParams.updated("retries", s"$i").asJava)
+    }
+    val toBeReleasedQueue = new ConcurrentLinkedQueue[CachedKafkaProducer]()
+
+    def acquire(i: Int): Unit = {
+      val producer = CachedKafkaProducer.acquire(kafkaParamsUniqueMap(i))
+      producer.kafkaProducer // materialize producer for the first time.
+      assert(!producer.isClosed, "Acquired producer cannot be closed.")
+      toBeReleasedQueue.add(producer)
+    }
+
+    def release(producer: CachedKafkaProducer): Unit = {
+      if (producer != null) {
+        CachedKafkaProducer.release(producer, true)
+        if (producer.getInUseCount > 0) {
+          assert(!producer.isClosed, "Should not close an inuse producer.")
+        }
+      }
+    }
+    val threadPool = Executors.newFixedThreadPool(numThreads)
+    try {
+      val futuresAcquire = (1 to 10 * numConcurrentProducers).map { i =>
+        threadPool.submit(new Runnable {
+          override def run(): Unit = {
+            acquire(i % numConcurrentProducers + 1)
+          }
+        })
+      }
+      val futuresRelease = (1 to 10 * numConcurrentProducers).map { i =>
+        val cachedKafkaProducer = toBeReleasedQueue.poll()
+        // 2x release should not corrupt the state of cache.
+        (1 to 2).map { j =>
+          threadPool.submit(new Runnable {
+            override def run(): Unit = {
+              release(cachedKafkaProducer)
+            }
+          })
+        }
+      }
+      futuresAcquire.foreach(_.get(1, TimeUnit.MINUTES))
+      futuresRelease.flatten.foreach(_.get(1, TimeUnit.MINUTES))
+    } finally {
+      threadPool.shutdown()
+    }
+  }
+
+  /*
+   * The following stress suite will cause frequent eviction of kafka producers from
+   * the guava cache. Since these producers remain in use, because they are used by
+   * multiple tasks, they stay in close queue till they are released finally. This test
+   * will cause new tasks to use fresh instance of kafka producers and as a result it
+   * simulates a stress situation, where multiple producers are requested from CachedKafkaProducer
+   * and at the same time there will be multiple releases. It is supposed to catch a race
+   * condition if any, due to multiple threads requesting and releasing producers.
+   */
+  test("Single source and multiple kafka sink with 2ms cache timeout.") {
```

Comment:
In case of deadlock this test never stopped on my machine.

Reply:
Is this problem resolved?

Reply:
Yeah, now the problem solved and runs fine. Thanks your efforts!